### PR TITLE
Add support for many compressed instructions.

### PR DESCRIPTION
**Pull request by user-0xcafe**
This PR adapts the package such that it can decompress all compressed instructions generated by a rv64imc GCC and provides iterators to iterate over instructions (both compressed and uncompressed). 
The iterators can provide either a stream of pc values or a list of decompressed and decoded instruction.

### CODE REVIEW

File: src/decompress/util.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+pub(super) fn build_btype(instruction_type: CbInstr, rs1: u16, imm: u16) -> u32 {
+    let mold = |imm: u16, rs1: u16, rs2: u16, funct3: u32, opcode: u32| -> u32 {
+        let rs1: u32 = rs1.into();
+        let rs2: u32 = rs2.into();
+
+        let imm: u32 = imm.permute(&[12, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 11]).into();
+
+        // FIXME: check if this is correct!
```
Comment (user-0xcafe) [inline src/decompress/util.rs:78]: This is already checked, will take it out.

File: src/decompress.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        _ => unreachable!(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{decode, types::*, Instruction::*};
+
+    // The bulk of the instructions were obtained by compiling selfie with
+    // an rv64imc enabled gcc compiler. Examples for case were created by hand.
```
Comment (user-0xcafe) [inline src/decompress.rs:63]: `case` should read `edge cases`

**Comment by mstarzinger** [comment]
Very nice! Well structured, good tests, I especially like how the iterators turned out. Thanks!
